jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -261,31 +268,408 @@ public InvocationType getInvocationType()
+                _contentQueue.poll();
+                if (c.isEof())
+                    _contentQueue.add(EOF);
+            }
+            return c;
+        }
+
+        @Override
+        public synchronized boolean demand()
+        {
+            if (!_contentQueue.isEmpty())
+                return true;
+            if (!_demanding)
+            {
+                _demanding = true;
+                IStream stream = getStream();
+                if (stream == null)
+                    _demanding = false;
+                else
+                    stream.demand(1);
+            }
+            return !_contentQueue.isEmpty();
+        }
+
+        @Override
+        public synchronized void onContent(HttpInput.Content content)
+        {
+            _demanding = false; // TODO maybe distinguish between upstream content and downstream errors
+            if (content.isEof() && content.isSpecial() && content.isEmpty() && _contentQueue.size() == 1 && _contentQueue.peek().isEof())
+            {
+                content.succeeded();
+                return;
+            }
+
+            // error content fails and replaces any other content
+            Throwable x = content.getError();
+            if (x != null)
+            {
+                _contentQueue.forEach(c -> c.failed(x));
+                _contentQueue.clear();
+            }
+
+            _contentQueue.offer(content);
+            if (_contentQueue.size() > 1)
+                throw new AssertionError("Queue should never be more than 1: " + _contentQueue);
+        }
+
+        @Override
+        public synchronized boolean onTimeout(Throwable failure)
+        {
+            if (_demanding && _contentQueue.isEmpty())
+            {
+                _demanding = false;
+                _contentQueue.add(new HttpInput.ErrorContent(failure));
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public synchronized void eof()
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("received EOF");
+
+            if (_contentQueue.size() != 1 || !_contentQueue.peek().isEof())
+                _contentQueue.add(EOF);
+        }
+
+        @Override
+        public synchronized void failContent(Throwable failure)
+        {
+            for (Iterator<HttpInput.Content> i = _contentQueue.iterator(); i.hasNext();)
+            {
+                HttpInput.Content c = i.next();
+                if (!c.isSpecial())
+                {
+                    i.remove();
+                    c.failed(failure);
+                }
+            }
+        }
+    }
+
+    private static final HttpInput.Content EOF = new HttpInput.EofContent();
+    private static final HttpInput.Content DEMANDING = new HttpInput.SpecialContent() {};
+
+    private class ContentDemanderAtomic implements ContentDemander
+    {
+        private final AtomicReference<HttpInput.Content> _content = new AtomicReference<>();
```

Comment:
We need a puml interaction diagram and/or state diagram for this state machine!

Reply:
I added a state diagram. I don't think an interaction diagram is going to help much as the code is relatively short and self contained.